pkg/registry/core/rest/storage_core.go

```diff
@@ -162,23 +162,39 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
 		return LegacyRESTStorage{}, genericapiserver.APIGroupInfo{}, err
 	}
 
+	var allocatorErr error
```

Comment:
it's not clear the passed-in function is called synchronously. suggest making AllocatorFactory return `(allocator.Interface, error)` instead, and NewAllocatorCIDRRange propagate the error

Reply:
Please take a look at allocator :
```
func NewAllocatorCIDRRange(cidr *net.IPNet, allocatorFactory allocator.AllocatorFactory) *Range {
...
	r.alloc = allocatorFactory(r.max, rangeSpec)
```
The passed-in function is executed synchronously.

Reply:
it happens to be today, but there is no guarantee of that behavior in the future

Reply:
The serviceClusterIPAllocator is used several lines down:
```
	serviceRest, serviceRestProxy := servicestore.NewREST(serviceRESTStorage, endpointsStorage, podStorage.Pod, serviceClusterIPAllocator, serviceNodePortAllocator, c.ProxyTransport)
```
Assuming serviceClusterIPAllocator points to a partially initialized instance at the time of assignment, I would expect some kind of synchronization before it is used in the servicestore.NewREST call above.
The instance serviceClusterIPAllocator points to is fully initialized when ipallocator.NewAllocatorCIDRRange returns.

Reply:
Then we should make that explicit by creating and passing in the instance, rather than a factory. Don't propagate an error to the outer function via a closure variable unless the invoked function guarantees the closure is invoked synchronously.

Reply:
I have adopted your initial suggestion where factory method return error.